staging/src/k8s.io/cloud-provider/controllers/service/controller.go

```diff
@@ -735,18 +735,29 @@ func (s *Controller) nodeSyncInternal(ctx context.Context, workers int) {
 	if !s.needFullSyncAndUnmark() {
 		// The set of nodes in the cluster hasn't changed, but we can retry
 		// updating any services that we failed to update last time around.
-		s.servicesToUpdate = s.updateLoadBalancerHosts(ctx, s.servicesToUpdate, workers)
+		// Besides, only latest Service specs should be applied.
+		var servicesToUpdate []*v1.Service
+		for key := range s.servicesToUpdate {
+			cachedService, exist := s.cache.get(key)
```

Comment:
The original issue talks about stale spec but this seems to only be accounting for deleted Services, is that expecdted here?

Reply:
In other words, should we actually check spec or is checking existencen enough?

Reply:
I think specs in cache are the source of truth so no need to check specs. Instead, always use `cachedService` to update.

Reply:
Yes, if it is a Service with stale spec, then on L746, it will be appended to `servicesToUpdate` list with correct spec from cache.

service to update -> not exist in cache -> skip
                             -> exist in cache -> update with correct spec from cache

Reply:
I see now, this is just reconstructing servicesToUpdate with latest cached objects

Reply:
So reading through the service controller code again, I'm not sure that calling `s.cache.get()` would fix this issue because it just fetches from an internal map:

https://github.com/kubernetes/kubernetes/blob/efb7f850fc623a11ec645e9f204934cc0712cf2f/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L492-L497

And I can't seem to find where `s.cache.set()` is called to update `m.serviceMap` :

https://github.com/kubernetes/kubernetes/blob/efb7f850fc623a11ec645e9f204934cc0712cf2f/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L510-L514

Perhaps that could be the actual issue here

Reply:
Please correct me if I'm wrong.

I think the internal map (cache) is the source of truth. Latest service spec is got in `syncService()` and the value is passed to `processServiceCreateOrUpdate()`.
https://github.com/kubernetes/kubernetes/blob/efb7f850fc623a11ec645e9f204934cc0712cf2f/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L870-L879

Indeed, `s.cache.set()` is never used because `cachedService.state = service` does the job in `processServiceCreateOrUpdate`.
`s.cache.set()` is useful if we want to update `s.serviceMap` with a new `cachedService` but here we just update the value with the existing `cachedService` pointer
https://github.com/kubernetes/kubernetes/blob/efb7f850fc623a11ec645e9f204934cc0712cf2f/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L345

Reply:
@andrewsykim do you have any other comments? Thank you!